node/auth/authenticator.go

```diff
@@ -0,0 +1,172 @@
+	// to the time when that cached authentication will expire.
+	authenticatedDispersers *lru.Cache[string, time.Time]
+
+	// authenticationTimeoutDuration is the duration for which an auth is valid.
+	// If this is zero, then auth saving is disabled, and each request will be authenticated independently.
+	authenticationTimeoutDuration time.Duration
+}
+
+// NewRequestAuthenticator creates a new RequestAuthenticator.
+func NewRequestAuthenticator(
+	ctx context.Context,
+	chainReader core.Reader,
+	keyCacheSize int,
+	keyTimeoutDuration time.Duration,
+	authenticationTimeoutDuration time.Duration,
+	now time.Time) (RequestAuthenticator, error) {
+
+	keyCache, err := lru.New[uint32, *keyWithTimeout](keyCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create key cache: %w", err)
+	}
+
+	authenticatedDispersers, err := lru.New[string, time.Time](keyCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create authenticated dispersers cache: %w", err)
+	}
+
+	authenticator := &requestAuthenticator{
+		chainReader:                   chainReader,
+		keyCache:                      keyCache,
+		keyTimeoutDuration:            keyTimeoutDuration,
+		authenticatedDispersers:       authenticatedDispersers,
+		authenticationTimeoutDuration: authenticationTimeoutDuration,
+	}
+
+	err = authenticator.preloadCache(ctx, now)
+	if err != nil {
+		return nil, fmt.Errorf("failed to preload cache: %w", err)
+	}
+
+	return authenticator, nil
+}
+
+func (a *requestAuthenticator) preloadCache(ctx context.Context, now time.Time) error {
+	// this will need to be updated for decentralized dispersers
+	_, err := a.getDisperserKey(ctx, now, EigenLabsDisperserID)
+	if err != nil {
+		return fmt.Errorf("failed to get operator key: %w", err)
+	}
+
+	return nil
+}
+
+func (a *requestAuthenticator) AuthenticateStoreChunksRequest(
+	ctx context.Context,
+	origin string,
+	request *grpc.StoreChunksRequest,
+	now time.Time) error {
+
+	if a.isAuthenticationStillValid(now, origin) {
+		// We've recently authenticated this client. Do not authenticate again for a while.
+		return nil
+	}
+
+	key, err := a.getDisperserKey(ctx, now, request.DisperserID)
+	if err != nil {
+		return fmt.Errorf("failed to get operator key: %w", err)
+	}
+
+	err = VerifyStoreChunksRequest(*key, request)
+	if err != nil {
+		return fmt.Errorf("failed to verify request: %w", err)
+	}
+
+	a.saveAuthenticationResult(now, origin)
+	return nil
+}
+
+// getDisperserKey returns the public key of the operator with the given ID, caching the result.
+func (a *requestAuthenticator) getDisperserKey(
+	ctx context.Context,
+	now time.Time,
+	disperserID uint32) (*gethcommon.Address, error) {
+	key, ok := a.keyCache.Get(disperserID)
+	if ok {
+		expirationTime := key.expiration
+		if now.Before(expirationTime) {
+			return &key.key, nil
+		}
+	}
+
+	address, err := a.chainReader.GetDisperserAddress(ctx, disperserID)
```

Comment:
I'd cache this ID to address mapping to save RPCs

Reply:
This is cached (see `keyCache *lru.Cache[uint32, *keyWithTimeout]`). By design, we re-fetch the same key again after a timeout just in case the key was changed. But not for each RPC.